test/generators/gen_link/gen_link.ml

```diff
@@ -0,0 +1,235 @@
+open Sexplib0.Sexp
+open Printf
+
+type sexp = t = Atom of string | List of sexp list
+
+let cu_target_rule path ext =
+  let open Fpath in
+  let basename = basename (rem_ext path) in
+  List
+    [
+      Atom "rule";
+      List [ Atom "target"; Atom (basename ^ ext) ];
+      List [ Atom "deps"; Atom (to_string path) ];
+      List
+        [
+          Atom "action";
+          List
+            [
+              Atom "run";
+              Atom "ocamlc";
+              Atom "-c";
+              Atom "-bin-annot";
+              Atom "-o";
+              Atom "%{target}";
+              Atom "%{deps}";
+            ];
+        ];
+    ]
+
+let odoc_target_rule path ext =
+  let open Fpath in
+  let basename = basename (rem_ext path) in
+  List
+    [
+      Atom "rule";
+      List [ Atom "target"; Atom (basename ^ ".odoc") ];
+      List [ Atom "deps"; Atom (basename ^ ext) ];
+      List
+        [
+          Atom "action";
+          List
+            [
+              Atom "run";
+              Atom "odoc";
+              Atom "compile";
+              Atom "--pkg";
+              Atom "test";
+              Atom "-o";
+              Atom "%{target}";
+              Atom "%{deps}";
+            ];
+        ];
+    ]
+
+let odocl_target_rule path =
+  let open Fpath in
+  let basename = basename (rem_ext path) in
+  List
+    [
+      Atom "rule";
+      List [ Atom "target"; Atom (basename ^ ".odocl") ];
+      List [ Atom "deps"; Atom (basename ^ ".odoc") ];
+      List
+        [
+          Atom "action";
+          List
+            [
+              Atom "run";
+              Atom "odoc";
+              Atom "link";
+              Atom "-o";
+              Atom "%{target}";
+              Atom "%{deps}";
+            ];
+        ];
+    ]
+
+let file_rule path ext =
+  let open Printf in
+  printf "%s\n\n" (to_string_hum (cu_target_rule path ext));
+  printf "%s\n\n" (to_string_hum (odoc_target_rule path ext));
+  printf "%s\n\n" (to_string_hum (odocl_target_rule path))
```

Comment:
Instead of using `printf` and `to_string_hum`, I suggest returning a `sexp list` and propagate it down to the main (`let ()`), where everything can be written at once using:
```ocaml
let () =
  let paths = read_file_from_dir "cases" in
  let stanzas = gen_rules paths in
  List.iter (Sexplib0.Sexp.pp Format.std_formatter) stanzas
```

Reply:
done